src/db-api.js

```diff
@@ -23,34 +31,53 @@ const unspentAddresses = (db: Pool) => async (): Promise<ResultSet> =>
     rowMode: 'array',
   })
 
+const utxoQuery = `SELECT 
+  TRIM(LEADING '\\x' from tx.hash::text) AS "tx_hash", tx_out.index AS "tx_index",
+  tx_out.address AS "receiver", tx_out.value AS "amount", tx.block::INTEGER as "block_num"
+FROM tx
+INNER JOIN tx_out ON tx.id = tx_out.tx_id
+WHERE NOT EXISTS (SELECT true
+  FROM tx_in
+  WHERE (tx_out.tx_id = tx_in.tx_out_id) AND (tx_out.index = tx_in.tx_out_index)
+) AND (tx_out.address = ANY($1))`
+
 /**
  * Queries UTXO table looking for unspents for given addresses
  *
  * @param {Db Object} db
  * @param {Array<Address>} addresses
  */
 const utxoForAddresses = (db: Pool) => async (addresses: Array<string>) =>
-  db.query('SELECT * FROM "utxos" WHERE receiver = ANY($1)', [addresses])
+  db.query({
+    text: utxoQuery,
+    values: [addresses],
+  })
 
 const utxoSumForAddresses = (db: Pool) => async (addresses: Array<string>) =>
-  db.query('SELECT SUM(amount) FROM "utxos" WHERE receiver = ANY($1)', [
-    addresses,
-  ])
-
-// Cached queries
-const txHistoryQuery = (limit: number) => `
-  SELECT *
-  FROM "txs"
-  LEFT JOIN (SELECT * from "bestblock" LIMIT 1) f ON true
-  WHERE 
-    hash = ANY (
-      SELECT tx_hash 
-      FROM "tx_addresses"
-      where address = ANY ($1)
-    )
-    AND last_update >= $2
-  ORDER BY last_update ASC
-  LIMIT ${limit}
+  db.query(`SELECT SUM(amount) FROM (${utxoQuery}) as utxo_table`, [addresses])
+
+const txHistory = (limit: number) => `
+  SELECT txs.id, txs.hash, txs.block_no, txs.blockHash, txs.block_index as tx_ordinal, txs.time, tx_body.body::text from (
+      SELECT
+        tx.id, tx.hash::text, block.block_no, block.hash::text as blockHash, block.time, tx.block_index
+        FROM block 
+        INNER JOIN tx ON block.id = tx.block 
+        INNER JOIN tx_out ON tx.id = tx_out.tx_id
+        WHERE tx_out.address = ANY($1)
+          AND block.time >= $2
+    UNION
+      SELECT DISTINCT 
+        tx.id, tx.hash::text, block.block_no, block.hash::text as blockHash, block.time, tx.block_index
+        FROM block 
+        INNER JOIN tx ON block.id = tx.block 
+        INNER JOIN tx_in ON tx.id = tx_in.tx_in_id 
+        INNER JOIN tx_out ON (tx_in.tx_out_id = tx_out.tx_id) AND (tx_in.tx_out_index = tx_out.index)
+        WHERE tx_out.address = ANY($1)
+          AND block.time >= $2
+    ORDER BY time ASC
+    LIMIT ${limit}
+  ) AS txs
+  JOIN tx_body ON txs.hash = tx_body.hash::text
```

Comment:
I tested locally and this join makes the query slow and CPU-hungry. However, I managed to speed the query up by rewriting it like this:

```
SELECT txs.id, txs.hash, txs.block_no, txs.blockHash, txs.block_index as tx_ordinal, txs.time, txs.body::text from (
       SELECT                                                                                                              
         tx.id, tx.hash::text, block.block_no, block.hash::text as blockHash, block.time, tx.block_index, tx_body.body::text                  
         FROM block                                                                                                        
         INNER JOIN tx ON block.id = tx.block                                                                              
         INNER JOIN tx_out ON tx.id = tx_out.tx_id
         JOIN  tx_body ON tx.hash = tx_body.hash                                                                  
         WHERE tx_out.address = ANY('{DdzFFzCqrhsk1hUcFMxLQ1fBJ7qSLvFGytti4a3HxTLtZHzDfe7F1GqxvVLUUpz1i9GkaBGUXgGudBNzehvB6WZAo2M6VFEL7yNcovSh}')                                                                                    
           AND block.time >= $2                                                                                  
     UNION                                                                                                                 
       SELECT DISTINCT                                                                                                     
         tx.id, tx.hash::text, block.block_no, block.hash::text as blockHash, block.time, tx.block_index, tx_body.body::text              
         FROM block                                                                                                        
         INNER JOIN tx ON block.id = tx.block                                                                              
         INNER JOIN tx_in ON tx.id = tx_in.tx_in_id                                                                        
         INNER JOIN tx_out ON (tx_in.tx_out_id = tx_out.tx_id) AND (tx_in.tx_out_index = tx_out.index)    
         JOIN  tx_body ON tx.hash = tx_body.hash                  
         WHERE tx_out.address = ANY($1)                                                                                                                                                                       
           AND block.time >= $2                                                                                      
     ORDER BY time ASC                                                                                                     
     LIMIT 20                                                                                                              
   ) AS txs
```

The likeliest explanation is that it is joining the hash of the "txs" temporary table which does not have an index.

Reply:
Also, you will need to add an index on block time for the query to be efficient:

`CREATE INDEX idx_block_time ON public.block USING btree (time);`

Reply:
sample query to replicate the performance issues:

```
SELECT txs.id, txs.hash, txs.block_no, txs.blockHash, txs.block_index as tx_ordinal, txs.time, tx_body.body::text from (
       SELECT                                                                                                              
         tx.id, tx.hash::text, block.block_no, block.hash::text as blockHash, block.time, tx.block_index                   
         FROM block                                                                                                        
         INNER JOIN tx ON block.id = tx.block                                                                              
         INNER JOIN tx_out ON tx.id = tx_out.tx_id                                                                         
         WHERE tx_out.address = ANY('{DdzFFzCqrhsk1hUcFMxLQ1fBJ7qSLvFGytti4a3HxTLtZHzDfe7F1GqxvVLUUpz1i9GkaBGUXgGudBNzehvB6WZAo2M6VFEL7yNcovSh}')                      
                                                              
           AND block.time >= '1990-01-01'                                                                                       
     UNION                                                                                                                 
       SELECT DISTINCT                                                                                                     
         tx.id, tx.hash::text, block.block_no, block.hash::text as blockHash, block.time, tx.block_index                   
         FROM block                                                                                                        
         INNER JOIN tx ON block.id = tx.block                                                                              
         INNER JOIN tx_in ON tx.id = tx_in.tx_in_id                                                                        
         INNER JOIN tx_out ON (tx_in.tx_out_id = tx_out.tx_id) AND (tx_in.tx_out_index = tx_out.index)                     
         WHERE tx_out.address = ANY('{DdzFFzCqrhsk1hUcFMxLQ1fBJ7qSLvFGytti4a3HxTLtZHzDfe7F1GqxvVLUUpz1i9GkaBGUXgGudBNzehvB6WZAo2M6VFEL7yNcovSh}')                      
                                                                                                                                                 
           AND block.time >= '1990-01-01'                                                                                         
     LIMIT 20                                                                                                              
   ) AS txs                                                                                                                
   JOIN tx_body ON txs.hash = tx_body.hash::text   ;
```